Feat. JetStream Support for Source

somratdutta opened the pull request:
We have added support for NATS JetStream for Source.
Also added is a test-case for the Source class.
We are open to feedback and actively working on to improve the current state of this project.

scottf commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+        return batchSize;
+    }
+
+    public boolean isCumulativeAck() {
```
What is this?

somratdutta commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+        return batchSize;
+    }
+
+    public boolean isCumulativeAck() {
```
sorry this was used in our own usecase, will be removed.

somratdutta replied:
It was mistakenly pushed, we have removed this now.
Thanks for noticing.

somratdutta commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+    private final String consumerName;
+    private final int batchSize;
+
+    private NATSConsumerConfig(Builder builder) {
```
Previously, in our use-case we were using this as a wrapper over ConsumerConfiguration class, simply as ConsumerConfiguration class isn't Serializable and Flink needs all fields to be serialized.
Either, we can make the ConsumerConfiguration class serializable or use a property based function (like done in Options).

somratdutta replied:
Yes, I agree, we shouldn't be using so many options.
Further plan is to send a PR in nats-client library to make it Serialisable.

somratdutta replied:
Since, the PR to make certain classes Serializable was rejected, I don't think we have other options.

scottf replied:
I see this now, we will have to make something.

scottf replied:
Here's what I'm doing. I'm changing the ConsumerConfiguration builder to accept json. That's just step one. This allows us to make a simple class to use:

```

public static class SerializableConsumerConfiguration implements Serializable {
    private static final long serialVersionUID = 1L;

    private String json;
    private transient ConsumerConfiguration cc;

    public SerializableConsumerConfiguration() {
        cc(ConsumerConfiguration.builder().build());
    }

    public SerializableConsumerConfiguration(ConsumerConfiguration cc) {
        cc(cc);
    }

    public SerializableConsumerConfiguration cc(ConsumerConfiguration cc) {
        this.cc = cc;
        json = cc.toJson();
        return this;
    }

    public String getJson() {
        return json;
    }

    public ConsumerConfiguration getConsumerConfiguration() {
        return cc;
    }

    private void writeObject(java.io.ObjectOutputStream out) throws IOException {
        out.defaultWriteObject();
    }

    private void readObject(java.io.ObjectInputStream in) throws IOException, ClassNotFoundException {
        in.defaultReadObject();
        cc = ConsumerConfiguration.builder().json(json).build();
    }
}
```

scottf replied:
I'm adding SerializableConsumerConfiguration directly to JNATS in the helpers section. Please review: https://github.com/nats-io/nats.java/pull/1156

somratdutta replied:
that's great @scottf, will review

somratdutta commented on src/main/java/io/synadia/flink/source/NATSJetstreamSource.java:
```diff
@@ -0,0 +1,98 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NATSJetstreamSource<T> implements Source<T, NatsSubjectSplit, Collection<NatsSubjectSplit>>, ResultTypeQueryable<T> {
```
Thanks, will make the changes accordingly.

somratdutta commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+package io.synadia.flink.source;
```
Can you please guide which license?
There's two as of now which is confusing me.
One is the Apache 2.0 one and other is Synadia one.

somratdutta commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+            return new NATSConsumerConfig(this);
+        }
+    }
+}
```
Added.

somratdutta commented on src/main/java/io/synadia/flink/source/NATSConsumerConfig.java:
```diff
@@ -0,0 +1,48 @@
+package io.synadia.flink.source;
```
Okay, refactored to `Nats`

somratdutta replied:
I totally agree with the idea of repackaging.
For an easy review of this PR, I will make those changes in the next PR as it would involve reviewing a lot of file changes.

shiv4289 replied:
Alternatively we could put it in a different commit if @scottf is fine to review by picking out commits individually.

somratdutta commented on src/main/java/io/synadia/flink/source/NATSJetstreamSource.java:
```diff
@@ -0,0 +1,98 @@
+    private final String id;
+
+    // Package-private constructor to ensure usage of the Builder for object creation
+    NATSJetstreamSource(DeserializationSchema<T> deserializationSchema, ConnectionFactory connectionFactory, String natsSubject, NATSConsumerConfig config) {
```
Flink has inbuilt implementation for Deserializer and Serializers.
In our application code, we are heavily invested in using Flink's native Deserializer (& Serializers) as our current Pulsar-Flink connector implements it.  In my humble opinion, this would be the same for other folks moving from another messaging framework. Thoughts?

scottf replied:
I don't think so. Consider the PayloadDeserializer interface
```
OutputT getObject(String subject, byte[] input, Headers headers);
```
versus the primary methods in the DeserializationSchema
```
T deserialize(byte[] var1) throws IOException;
```
The Nats message is more than just the data byte array. But there is your code could implement both interfaces and delegate the getObject implementation to call deserialize.

somratdnutanix commented on src/main/java/io/synadia/flink/source/js/NatsConsumerConfig.java:
```diff
@@ -0,0 +1,62 @@
+package io.synadia.flink.source.js;
+import java.io.Serializable;
+
+public class NatsConsumerConfig implements Serializable {
```
Noted, 
here's a list of suggested name for the above class, recommendation needed:

- NatsConsumeOptions
- NatsBehaviorSettings
- NatsRuntimeParameters
- NatsOperationalConfig
- NatsProcessingOptions
- NatsBehaviorConfig
- NatsDynamicConfig
- NatsPerformanceSettings
- NatsConsumeBehavior
- NatsConsumeParameters

Going for NatsConsumeOptions for now.

vu-hoang-phan-form3 commented on src/main/java/io/synadia/flink/source/NatsJetStreamSource.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public SplitEnumerator<NatsSubjectSplit, Collection<NatsSubjectSplit>> createEnumerator(
```
If I understand correctly, a split will be a NATS subject. I'm curious about whether Flink will spawn multiple source reader in parallel for a single split?

I'm asking because with Kafka, a split is a partition, and apparently only one consumer will read that partition at a time. NATS is different that multiple consumers can read from the same subject. If I need multiple readers to read a very big subject, how Flink will handle that?

scottf replied:
Yes the subject is a split. Should only be one reader per split (per subject)

vu-hoang-phan-form3 replied:
If one has only one huge subject, one reader might not be enough to handle the rate. Usually we can spawn multiple nats client to read from the same subject but with this pattern there can only be one I believe.

scottf replied:
I think we could try to figure out how to use multiple instances of the same pull consumer (works the same as push delivery group). So when that option is configured we would have to know the number of consumer instances and build the splits accordingly.

sourabhaggrawal replied:
I agree that we should build support for multiple consumers reading from one subject. I will work on it. 
However, for this PR, would prefer to limit it to one subject one consumer. 
We can create an issue in this project as requirement for record keeping.

tarun-annapareddy replied:
@vu-hoang-phan-form3 @scottf Why can't we have a basic NATS Jetstream Flink connector that reads from the stream with one consumer in this base PR? Multiple consumers for the stream(one for each subject) need new subject discovery, we should take both as improvements together.

scottf replied:
The intention is to have both. It will have to be some sort of configuration. This needs to be done for the non jetstream as well.

sourabhaggrawal replied:
Hello @scottf, just to clear if I understand it right, you are suggesting that we should have :
1. one consumer for every subject under a regex subject like (ORDER.*) ?
2.one consumer for every subject from list of subjects that user will provide ?
3.multiple copies of one consumer to increase the read throughput, if subject is big ?

If I have understood it correctly you are suggesting to have multiple copies of one consumer, right ?

scottf replied:
The original implementation made one split per subject filter. I'm suggesting that we need some sort of configuration hint that tell us to make N splits for a single subject filter. I suppose that could be accomplished by allowing the subject filter multiple times in the input list.
Also, I would not characterize a subject like "order.*" as a regex, it's a wildcard. As presented, it is simply a single subject filter to a consumer.

sourabhaggrawal replied:
Yes order.* is one subject and thats how I see it too. Now I have got more clarity on this multiple consumer part which I was confusing myself with some other usecase.
Let me see it once again how we can fit creating multiple instance of same consumer for one subject.

vu-hoang-phan-form3 replied:
What if we don't need to have split at all? NATS supports multiple readers out-of-the-box already. So a NATS source just need to create multiple readers with the same config (same stream, same subject) and they will work in parallel without further config.

souravagrawal replied:
Hi @vu-hoang-phan-form3  It does make sense to ignore split if we do not want to use checkpointing, which is the current implementation, and we are not making use of the Split. 
But to implement the checkpointing we need Split, it holds the data which is fetched from source and same data needs to be provided back to Flink during snapshot.
I have changed the code to implement checkpoint here 
https://github.com/souravagrawal/flink-connector-nats/commit/9c7c54546c8edff0e8a3482a5d3787cd3be6512c

Please go through with it once, I plan to merge it in somrat's fork soon.

vu-hoang-phan-form3 commented on src/main/java/io/synadia/flink/source/NatsJetStreamSourceReader.java:
```diff
@@ -0,0 +1,157 @@
+    @Override
+    public InputStatus pollNext(ReaderOutput<OutputT> output) throws Exception {
+        List<Message> messages =
+                ((JetStreamSubscription) subscription).fetch(config.getBatchSize(), Duration.ofSeconds(5));
```
`maxWait` here is hard-coded to 5s. I think it could be configurable?

scottf replied:
// TODO

sourabhaggrawal replied:
Agreed, @somratdutta please make it configurable.

somratdutta replied:
sure will do it

vu-hoang-phan-form3 commented on src/main/java/io/synadia/flink/source/NatsJetStreamSourceReader.java:
```diff
@@ -0,0 +1,157 @@
+                ((JetStreamSubscription) subscription).fetch(config.getBatchSize(), Duration.ofSeconds(5));
+        for (int i = 0; i < messages.size(); i++) {
+            Message message = messages.get(i);
+            boolean ackMessageFlag = (i == messages.size() - 1);
+            processMessage(output, message, ackMessageFlag);
```
If a message is acknowledged but checkpointing fails, we lose the message.
I think message ack should be inside `notifyCheckpointComplete` method instead. Either that or inside `snapshotState`.

somratdutta replied:
Sure, will resolve this.

sourabhaggrawal replied:
Hi @vu-hoang-phan-form3, message is acknowledged in pollNext as we read, if for any reason checkpointing is failed after acknowledging the job for the message is anyways done, there is a chance of getting same message again if for some reason connector gets rebooted during message delivery and before it could ack the message, consumer will start redelivering message from where it stopped. 
If you can help me understand how do we lose the message here if checkpointing is failed ?

shiv4289 replied:
@vu-hoang-phan-form3 you mean the processed data for the message might get lost?

vu-hoang-phan-form3 replied:
@shiv4289 yes
@sourabhaggrawal After `processMessage` here, the message is acknowledge and goes down the pipeline. Then if the next checkpoint fails, that same message will be lost. 

There is indeed a chance of redelivering the same message twice if it's acknowledged in `notifyCheckpointComplete`. That method is best-effort anyways. But it's not a problem, deduplication can be implemented elsewhere. Plus, NATS is at-least-once delivery, so duplication can always happen.

sourabhaggrawal replied:
@vu-hoang-phan-form3 @scottf can you please take a look at it once again, we have implemented checkpointing in this [PR](https://github.com/somratdutta/flink-connector-nats/pull/20/files)

somratdutta commented:
Hi @scottf,
I have made some changes as per the latest discussion.
You can see the commit [here](https://github.com/synadia-io/flink-connector-nats/pull/4/commits/b3cc3048e1cdff4bef5bd48ffa42faeae4b0f703).
I also agree on your PR for supporting [Serializable Consumer Configuration](https://github.com/nats-io/nats.java/pull/1156), I would refactor the Jetstream connector code once its merged.

somratdutta commented:
Updates in this [commit](https://github.com/synadia-io/flink-connector-nats/pull/4/commits/4a40c0f36b7618ea788c8f6699e32674a3f790c9) :

- Removed NatsConsumeOptions and updated configurations in NatsJetStreamSource and related classes:
  - Replaced NatsConsumeOptions with SerializableConsumerConfiguration for more efficient JSON serialization.
  - Updated NatsJetStreamSource, NatsJetStreamSourceBuilder, and related classes to use SerializableConsumerConfiguration.
  - Ensured proper handling of ConsumerConfiguration instances through SerializableConsumerConfiguration.

- Fixed Method Visibility in ExampleErrorListener to Match Interface Requirements:
  - Adjusted method visibility in ExampleErrorListener to comply with the interface requirements.
  - Ensured all interface methods are properly implemented with the correct access modifiers.

- Refactored NatsJetStreamSourceTest:
  - Modified test cases to use SerializableConsumerConfiguration instead of NatsConsumeOptions.
  - Ensured proper thread interruption handling and Flink environment cleanup.
  - Added explicit null checks and descriptive comments for better code readability.

Note :
This refactoring is done based on this [PR](https://github.com/nats-io/nats.java/pull/1156) which adds Serialisable wrappers. It hasn't been released yet, so the jnats version is bumped for next release.

tkeller-moxe commented:
Hey our team is interested in this!
Do we know the status of this PR?
